rfq/negotiator.go

```diff
@@ -293,6 +300,67 @@ func (n *Negotiator) HandleIncomingBuyRequest(
 	return nil
 }
 
+// HandleIncomingSellRequest handles an incoming asset sell quote request.
+// This is a request by our peer to sell an asset to us.
+func (n *Negotiator) HandleIncomingSellRequest(
+	request rfqmsg.SellRequest) error {
+
+	// TODO(ffranr): Ensure that we have a suitable buy offer for the asset
+	//  that our peer is trying to sell to us.
+
+	// Define a thread safe helper function for adding outgoing message to
+	// the outgoing messages channel.
+	sendOutgoingMsg := func(msg rfqmsg.OutgoingMsg) {
+		sendSuccess := fn.SendOrQuit(
+			n.cfg.OutgoingMessages, msg, n.Quit,
+		)
+		if !sendSuccess {
+			err := fmt.Errorf("negotiator failed to add message "+
+				"to the outgoing messages channel (msg=%v)",
+				msg)
+			n.cfg.ErrChan <- err
+		}
+	}
+
+	// Query the price oracle asynchronously using a separate goroutine.
+	// The price oracle might be an external service, responses could be
+	// delayed.
+	n.Wg.Add(1)
+	go func() {
+		defer n.Wg.Done()
+
+		// Query the price oracle for a bid price. This is the price we
+		// are willing to pay for the asset that our peer is trying to
+		// sell to us.
+		bidPrice, bidExpiry, err := n.queryBidFromPriceOracle(
+			request.Peer, request.AssetID, request.AssetGroupKey,
+			request.AssetAmount,
+		)
+		if err != nil {
+			// Send a reject message to the peer.
+			msg := rfqmsg.NewReject(
+				request.Peer, request.ID,
+				rfqmsg.ErrUnknownReject,
+			)
+			sendOutgoingMsg(msg)
+
+			// Add an error to the error channel and return.
+			err = fmt.Errorf("failed to query ask price from "+
+				"oracle: %w", err)
+			n.cfg.ErrChan <- err
+			return
+		}
+
+		// Construct and send a sell accept message.
+		msg := rfqmsg.NewSellAcceptFromRequest(
+			request, bidPrice, bidExpiry,
+		)
+		sendOutgoingMsg(msg)
+	}()
+
+	return nil
```

Comment:
There's a TODO in this function that will end up with an error return code path:
```
// TODO(ffranr): Ensure that we have a suitable buy offer for the asset
//  that our peer is trying to sell to us.
```

It will end up looking like `HandleIncomingBuyRequest`.

Reply:
Makes sense :+1: